[Feature] Add JSON_SET function

arvindKandpal-ksolves opened the pull request:
## Why I'm doing:

StarRocks currently lacks the `JSON_SET()` function, which exists in MySQL and is widely used for manipulating JSON data. This function is essential for "Upsert" operations (Update if exists, Insert if not) on JSON documents.

## What I'm doing:

This PR implements the `JSON_SET` function following MySQL compatibility.

### Function Signature
`JSON_SET(json_doc, path, val[, path, val] ...)`

### Implementation Details
1.  **Frontend:** Registered the function in `FunctionSet.java` and `functions.py`. Added support in `PruneSubfieldRule.java`.
2.  **Backend (C++):** Implemented recursive logic using `arangodb::velocypack::Builder` to traverse and rebuild the JSON document.
    - Handles **Object keys**: Inserts new keys or updates existing ones.
    - Handles **Array indices**: Updates specific indices (e.g., `$.arr[1]`) or appends if the index is out of bounds.
3.  **Documentation:** Added user documentation in `json_set.md`.
4.  **Testing:** Added SQL regression tests covering edge cases (nested paths, type handling, array modification).

### Verification
Verified locally with manual queries.
**Example: Array Index Update**
Query: `SELECT json_set(data, '$.arr[1]', 99) FROM json_set_test_table;`
Result: `{"arr": [10, 99, 30]}` (Correctly updates index 1 instead of replacing the array)

<img width="908" height="410" alt="Screenshot from 2025-12-02 09-59-05" src="https://github.com/user-attachments/assets/c4eea71a-2c91-48eb-867a-5e1b640210be" />


Fixes #66120

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
- [x] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.0
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds `json_set` to insert/update JSON values by path, with BE implementation, FE registration, documentation, and comprehensive tests.
> 
> - **Backend (C++)**:
>   - Implement `JsonFunctions::json_set` with recursive writers (`_json_set_recursive`, `_json_set_one_path`) to upsert values by `JSONPath`, including array index updates/appends and root replacement.
>   - Integrate into `json_functions.cpp/h` and handle null/argument validation.
> - **Frontend/Registration**:
>   - Register `JSON_SET` in `fe/src/.../FunctionSet.java`.
>   - Map function in `gensrc/script/functions.py` (`id 110025`, signature `JSON, JSON, ...` -> `JsonFunctions::json_set`).
> - **Docs**:
>   - Add `docs/.../json_set.md` covering syntax, parameters, return, examples, and notes.
> - **Tests**:
>   - Add C++ unit tests (`JsonSetTestFixture`) with extensive cases.
>   - Add SQL regression tests `test/sql/test_json/{T,R}/test_json_set`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 11a99facb8fc894d5acb2e1c52e408fc15034bfe. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

arvindKandpal-ksolves commented:
Hi @murphyatwork ,

I have implemented the `JSON_SET` function as requested in #66120.

This implementation follows MySQL compatibility and correctly handles the "Upsert" logic (insert if the path doesn't exist, update if it does).

I have verified the changes locally against various edge cases, including:
- Nested object modifications.
- Array index updates (ensuring specific indices are updated without replacing the entire array).
- Handling distinct data types and null arguments.

I have also included the comprehensive SQL regression tests and the user documentation (`json_set.md`).

The PR is ready for review. Thanks!

murphyatwork commented on be/src/exprs/json_functions.cpp:
```diff
@@ -1560,6 +1560,215 @@ static bool json_slice_contains(const arangodb::velocypack::Slice& target,
+        
+        for (auto it : vpack::ObjectIterator(slice)) {
+            std::string current_key = it.key.copyString();
+            if (current_key == target_key) {
```
The `vpack::Slice` includes a lookup key method for efficient binary searching.

arvindKandpal-ksolves commented:
Thanks @murphyatwork and @alvin-celerdata for the review!

I am working on the fixes:
1. Addressing the unsafe `down_cast` issue flagged by the bot (adding proper type checks for array selectors).
2. Optimizing string comparison using `isEqualString` as suggested to avoid memory allocation.
3. Adding a note in the documentation regarding wildcard limitations for clarity.

I will push the updates shortly.

alvin-celerdata commented:
@cursor review

alvin-celerdata commented:
@cursor review

alvin-celerdata commented:
@cursor review

alvin-celerdata commented:
@cursor review

arvindKandpal-ksolves committed:
[Feature] Add JSON_SET function

This PR implements the JSON_SET function, which allows inserting or updating data in a JSON document at specified paths (Upsert behavior), following MySQL compatibility.

Summary of changes:
- Backend (C++): Implemented recursive logic using vpack::Builder for object/array modification. Added safety checks for array selectors and optimizations for string comparison.
- Frontend (Java): Registered function signature and added optimizer support.
- Documentation: Added user guide with examples and usage notes.
- Testing: Added comprehensive SQL regression tests and Backend unit tests covering edge cases, nested paths, and null handling.

Fixes #66120

Signed-off-by: arvindksi274-ksolves <arvind.kandpal@ksolves.com>

arvindKandpal-ksolves committed:
test: add comprehensive test cases to improve backend coverage

Signed-off-by: arvindksi274-ksolves <arvind.kandpal@ksolves.com>

murphyatwork commented:
@arvindKandpal-ksolves current failed test is `    2421 ms: /root/starrocks/be/ut_build_Release/test/starrocks_test LakeServiceTest.test_abort_txn2`
